src/ripple/app/paths/impl/BookStep.cpp

```diff
@@ -156,34 +162,269 @@ class BookStep : public StepImp<TIn, TOut, BookStep<TIn, TOut>>
 
     bool equal (Step const& rhs) const override;
 
+    // Iterate through the offers at the best quality in a book.
+    // Unfunded offers and bad offers are skipped (and returned).
+    // callback is called with the offer SLE, taker pays, taker gets.
+    // If callback returns false, don't process any more offers.
+    // Return the unfunded and bad offers and the number of offers consumed.
+    template <class Callback>
+    std::pair<boost::container::flat_set<uint256>, std::uint32_t>
+    forEachOffer (
+        PaymentSandbox& sb,
+        ApplyView& afView,
+        bool prevStepRedeems,
+        Callback& callback) const;
+
     void consumeOffer (PaymentSandbox& sb,
         TOffer<TIn, TOut>& offer,
         TAmounts<TIn, TOut> const& ofrAmt,
         TAmounts<TIn, TOut> const& stepAmt,
         TOut const& ownerGives) const;
+};
+
+//------------------------------------------------------------------------------
+
+// Flow is used in two different circumstances for transferring funds:
+//  o Payments, and
+//  o Offer crossing.
+// The rules for handling funds in these two cases are almost, but not
+// quite, the same.
+
+// Payment BookStep template class (not offer crossing).
+template<class TIn, class TOut>
+class BookPaymentStep
+    : public BookStep<TIn, TOut, BookPaymentStep<TIn, TOut>>
+{
+public:
+    using BookStep<TIn, TOut, BookPaymentStep<TIn, TOut>>::BookStep;
+    using BookStep<TIn, TOut, BookPaymentStep<TIn, TOut>>::qualityUpperBound;
+
+    // Never limit self cross quality on a payment.
+    bool limitSelfCrossQuality (AccountID const&, AccountID const&,
+        TOffer<TIn, TOut> const& offer, boost::optional<Quality>&,
+        FlowOfferStream<TIn, TOut>&, bool) const
+    {
+        return false;
+    }
+
+    // A payment can look at offers of any quality
+    bool checkQualityThreshold(TOffer<TIn, TOut> const& offer) const
+    {
+        return true;
+    }
+
+    // For a payment ofrInRate is always the same as trIn.
+    std::uint32_t getOfrInRate (
+        Step const*, TOffer<TIn, TOut> const&, std::uint32_t trIn) const
+    {
+        return trIn;
+    }
+
+    // For a payment ofrOutRate is always the same as trOut.
+    std::uint32_t getOfrOutRate (Step const*, TOffer<TIn, TOut> const&,
+        AccountID const&, std::uint32_t trOut) const
+    {
+        return trOut;
+    }
+
+    Quality
+    qualityUpperBound(ReadView const& v,
+        Quality const& ofrQ,
+        bool prevStepRedeems) const
+    {
+        // Charge the offer owner, not the sender
+        // Charge a fee even if the owner is the same as the issuer
+        // (the old code does not charge a fee)
+        // Calculate amount that goes to the taker and the amount charged the
+        // offer owner
+        auto rate = [&](AccountID const& id) -> std::uint32_t {
```

Comment:
Interesting.  The code we're looking at is currently inaccessible, so I can't write a unit test for it.  We're computing the `qualityUpperBound()` for a `BookPaymentStep`.  All calls to `qualityUpperBound()` are guarded by `if (offerCrossing)` here: https://github.com/ripple/rippled/pull/1884/files#diff-44471707b2d421abd91624323bb9f1c5R475.  So the payment version of `qualityUpperBound()` can never be called (today).

Therefore I have a commit with some untested but probably correct code for you to look over (it's pretty much exactly your code).